File: python/bc-lifehash/src/bc_lifehash/_change_grid.py
from __future__ import annotations

from ._grid import Grid


class ChangeGrid:
    """Tracks which cells and their neighborhoods changed."""

    def __init__(self, width: int, height: int) -> None:
        self.grid: Grid[bool] = Grid(width, height, False)

    def set_changed(self, px: int, py: int) -> None:
        w = self.grid.width
        h = self.grid.height
        for oy in range(-1, 2):
            for ox in range(-1, 2):
                nx = (ox + px) % w
                ny = (oy + py) % h
                self.grid.set_value(True, nx, ny)

File: go/bclifehash/change_grid.go
package bclifehash

// changeGrid tracks which cells (and their neighborhoods) have changed.
type changeGrid struct {
	grid grid[bool]
}

func newChangeGrid(width, height int) changeGrid {
	return changeGrid{grid: newGrid[bool](width, height)}
}

func (cg *changeGrid) setChanged(px, py int) {
	width := cg.grid.width
	height := cg.grid.height
	for oy := -1; oy <= 1; oy++ {
		for ox := -1; ox <= 1; ox++ {
			nx := circularIndex(ox+px, width)
			ny := circularIndex(oy+py, height)
			cg.grid.setValue(true, nx, ny)
		}
	}
}
